### PR TITLE
Support `chainerx::Take` indices other dtype than int64

### LINKED ISSUE #6483: Support int32 indices in chainerx::take

### PR DESCRIPTION
**Pull request by niboshi**
Fixes #6483

### CODE REVIEW

File: chainerx_cc/chainerx/routines/indexing.cc [hunk 1]
```diff
@@ -156,34 +156,33 @@ Array AddAt(const Array& a, const Array& indices, int8_t axis, const Array& b) {
...
-    // TODO(niboshi): Support other dtypes by casting
-    if (indices.dtype() != Dtype::kInt64) {
-        throw DtypeError(
-                std::string{"Only "} + GetDtypeName(Dtype::kInt64) + " is supported as indices, but given " +
-                GetDtypeName(indices.dtype()));
+    DtypeKind indices_kind = GetKind(indices.dtype());
+    if (!(indices_kind == DtypeKind::kInt || indices_kind == DtypeKind::kUInt)) {
+        throw DtypeError{"Dtype ", GetDtypeName(indices.dtype()), " cannot be used as an indices array."};
     }
+    const Array& indices_cast = indices.dtype() == Dtype::kInt64 ? indices : indices.AsType(Dtype::kInt64);
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/indexing.cc]: Should we let the device implementation decide whether this cast is needed or not?

**Comment by hvy** [comment]
How about writing Python tests instead of C++ ones (since C++ tests will be deleted by https://github.com/chainer/chainer/pull/6438)?

**Comment by niboshi** [comment]
>How about writing Python tests instead of C++ ones (since C++ tests will be deleted by #6438)?

I'll do that if #6438 is merged before this PR.

File: chainerx_cc/chainerx/routines/indexing.cc (hunk 1 above)
Comment (niboshi) [reply chainerx_cc/chainerx/routines/indexing.cc]: Fixed as suggested.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
Discussed offline to add a TODO comment about not always creating a copy.

**Comment by niboshi** [comment]
Implemented castless int32 indexing in CUDA.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
Can you check the CI failures? Maybe it's a nvcc compilation failure?

**Comment by niboshi** [comment]
Fixed.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by asi1024** [comment]
LGTM.

**Pull request merged**